components/iosnoop/src/biosnoop.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use bcc::perf_event::PerfMapBuilder;
+use bcc::{Kprobe, BPF};
+
+use core::sync::atomic::{AtomicBool, Ordering};
+use std::ptr;
+use std::sync::{Arc, Mutex};
+use std::thread::{Builder as ThreadBuilder, JoinHandle};
+use tikv_util::collections::HashMap;
+
+lazy_static! {
+    pub static ref IO_READ_BYTES_PER_PID: Mutex<HashMap<u32, u64>> = Mutex::new(HashMap::default());
+}
+
+pub struct IOContext {
+    tid: u32,
+    read_bytes: u64,
+}
+
+impl IOContext {
+    pub fn new() -> Self {
+        let tid = nix::unistd::gettid().as_raw() as u32;
+        IOContext {
+            tid,
+            read_bytes: *IO_READ_BYTES_PER_PID
+                .lock()
+                .unwrap()
+                .entry(tid)
+                .or_default(),
+        }
+    }
+
+    fn delta(&self) -> u64 {
+        *IO_READ_BYTES_PER_PID
+            .lock()
+            .unwrap()
+            .entry(self.tid)
+            .or_default()
+            - self.read_bytes
+    }
+}
+
+#[repr(C)]
+struct data_t {
+    pid: u32,
+    rwflag: u64,
+    len: u64,
+}
+
+pub struct IOSnooper {
+    handle: Option<JoinHandle<()>>,
+    runnable: Arc<AtomicBool>,
+}
+
+impl IOSnooper {
+    pub fn new() -> Self {
+        IOSnooper {
+            handle: None,
+            runnable: Arc::new(AtomicBool::new(true)),
+        }
+    }
+
+    pub fn start(&mut self) -> Result<(), String> {
+        let code = include_str!("biosnoop.c");
+        info!("My pid is {}", nix::unistd::getpid());
```

Comment:
Can we include stats only if the PID is the same? In case multiple TiKV processes is co-located on a single host.

Reply:
yes, that's what `##TGID##` does. In kernel level, tgid -> pid, pid -> tid.

Reply:
Yes, that's what `##TGID##` does. In kernel level: tgid -> pid, pid -> tid.